### ISSUE TITLE
OverloadMethodsDeclarationOrder: new property to allow few groups of overloaded methods

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#3770]
This is a summary of #3698 .
Successor for: #2652, #3235 .

https://checkstyle.org/config_coding.html#OverloadMethodsDeclarationOrder

property "ModifiersThatCreateGroup" = list of RegExp for modifiers.
It is ok to provide better name for property. 

Examples : 
`ModifiersThatCreateGroup` = "static" - all static methods should be separate (should cover request of this issue), 
"ModifiersThatCreateGroup" = "public, private, protected, package" - 4 groups, each visibility modifier in its own. There will be 'package' artificial visibility modifier.
"ModifiersThatCreateGroup" = "static final , public native, private static, protected abstract" - 4 groups it will keep method with "static final" , "public native", "private static", "protected abstract" is separate groups, modifiers are completely matching , order is matter

Examples:
```
// ModifiersThatCreateGroup = "static"
public class Test {
    void foo() {}
    void foo(String baz) {}
    void bar() {} // OK, bar group
    static void foo() {} // OK, as "static foo" is separate group
}

// ModifiersThatCreateGroup = "static"
public class Test {
    static private void foo() {  }
    void foo() {  } // OK
    void foo(String baz) {  } //OK
    void bar() {  }
    static void foo() {  } // WARN - "static foo" methods are not groupped
}


// ModifiersThatCreateGroup = "static"
public class Test {
    public void foo() {  } //OK
    public void bar(String baz) {  } //OK
    private void foo(String baz) {  } //VIOLATION private foo(String) should be next to foo()
    void bar() {  } // VIOLATION bar should be next to bar(String)
    static public void foo() {  } // OK, "static foo" is separate group.
}

// ModifiersThatCreateGroup = "static"
public class Test {
    public void foo() {  } //OK
    private void bar(String baz) {  } //OK
    private void foo(String baz) {  } //VIOLATION private foo(String) should be next to foo()
    void bar() {  } // VIOLATION bar should be next to bar(String)
    static protected final void foo() {  } // OK, static foo is it's own group. 
}

================

//ModifiersThatCreateGroup = "protected, private, package"
public class Test {
    private void foo() {  } //OK, 2nd group
    void bar() {  }
    public void foo(String baz) {  } //OK, "public foo" does not match special groups so it is default group
    public void foo(String baz) {  } //OK
    private static void foo(int i) {  } // VIOLATION, matches 2nd group, should be close to first method
    private final void foo(int i) {  } // VIOLATION, matches 2nd group, should be close to first method
}

//ModifiersThatCreateGroup = "protected, private, package"
public class Test {
    protected void foo() {  } //OK, matches 1st group
    void bar() {  }
    protected abstract void foo() {  } //VIOlATION, matches 1st group
    void foo(int i) {  } //OK, matches 3rd group
    public final void foo(int i) {  } // OK, "public final " does not match special groups so it is default group
}


================

//ModifiersThatCreateGroup = "public, public final, static"
public class Test {
    private void foo() {  } // default group
    void bar() {  } // defaut group
    static foo() {} // OK, 3rd group
    static final foo() {} // OK, matches to 3rd group
    final void foo(String baz) {  } //VIOLATION, matching to default group
    final public void foo(String baz) {  } //OK, 1st group, Attention: does not match second group, as strict order of modifiers is in regexp
    private void foo(int i) {  } // VIOLATION, default group, should be close to first method
    public final void foo(int i) {  } // OK, 2nd group
    abstract foo() {} //VIOLATION, matches to default, should be close to first method
}


//ModifiersThatCreateGroup = RegExp[] = "^public .*, (protected|package), private, ^static$"
public class Test {
    private void foo() {  } // 3rd group
    void bar() {  } // second group
    static foo() {} // OK, 4th group
    static final foo() {} // OK, matches to default group
    final void foo(String baz) {  } //OK, matching to default group
    public void foo(String baz) {  } //OK, first group
    private void foo(int i) {  } // VIOLATION, 3rd group, should be close to first method
    public final void foo(int i) {  } // VIOLATION, matches to first group
    abstract foo() {} //VIOLATION, matches to default
}
```

**Comment by ilanKeshet** [comment]
Hi @romani, everybody 
I've started working on #3770 
You can see the work already done at https://github.com/ilanKeshet/checkstyle/commits/i3770

**Comment by ilanKeshet** [comment]
@romani I have the implementation at a stage where it performs the required functionality with units tests. but it doesn't yet pass all of the checkstyle rules (though it does successfully pass most).


Is there a review process, does it only start after a pull request, what's the next step?

**Comment by rnveach** [comment]
@ilanKeshet The next process is a PR, but we usually don't start reviewing them until all CI passes.
We will provide assistance if you need help getting the CI to pass.
If your not ready for a PR, you can start a conversation in our google groups.

**Comment by romani** [comment]
@ilanKeshet , please fix violations it will ease a lot review on my side. As you succeeded "mvn verify" please create PR. It is not a rare case then PR is full of discussions and redoing and .... .

If you want to discuss something before PR, maillist is good choice, or here at the issue.

**Comment by ilanKeshet** [comment]
Thanks @rnveach @romani

**Comment by romani** [comment]
@ilanKeshet, do you still plan to provide fix ?
It is completely ok to call back your original desire to help. It will let others to make it happen.

**Comment by ilanKeshet** [comment]
Thank you for the reminder, 
I would like to complete what I've started, and would resume work on Friday

**Comment by romani** [comment]
issue is still not resolved, everyone is welcome to contribute.

**Comment by Vampire** [comment]
@ilanKeshet any plans on getting this done? :-)

**Comment by austinarbor** [comment]
@romani I'm not sure why @ilanKeshet stopped working on this, from what I can tell he was pretty close and all tests were passing. I'm surprised this hasn't been picked up / completed yet. I created a new fork and merged in master to his working branch and resolved the conflicts to the best I could. All the unit-tests are passing, but I am getting some failures on the xml validations. Do you know what needs to be done?

```
[ERROR] Tests run: 7, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.018 s <<< FAILURE! - in com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest
[ERROR] testAllCheckSections  Time elapsed: 0.053 s  <<< FAILURE!
com.google.common.truth.AssertionErrorWithFacts:
config_coding.xml section 'OverloadMethodsDeclarationOrder' could not find field modifierGroups
expected to be true
	at com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest.getFieldClass(XdocsPagesTest.java:1260)
	at com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest.validatePropertySectionPropertyEx(XdocsPagesTest.java:892)
	at com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest.validatePropertySectionProperties(XdocsPagesTest.java:861)
	at com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest.validatePropertySection(XdocsPagesTest.java:725)
	at com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest.validateCheckSection(XdocsPagesTest.java:583)
	at com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest.testAllCheckSections(XdocsPagesTest.java:500)
```

```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.629 s <<< FAILURE! - in com.puppycrawl.tools.checkstyle.internal.XdocsJavaDocsTest
[ERROR] testAllCheckSectionJavaDocs  Time elapsed: 0.629 s  <<< ERROR!
java.lang.NullPointerException
	at com.puppycrawl.tools.checkstyle.internal.XdocsJavaDocsTest.createPropertiesText(XdocsJavaDocsTest.java:311)
	at com.puppycrawl.tools.checkstyle.internal.XdocsJavaDocsTest.examineCheckSubSection(XdocsJavaDocsTest.java:222)
	at com.puppycrawl.tools.checkstyle.internal.XdocsJavaDocsTest.examineCheckSectionChildren(XdocsJavaDocsTest.java:208)
	at com.puppycrawl.tools.checkstyle.internal.XdocsJavaDocsTest.examineCheckSection(XdocsJavaDocsTest.java:187)
	at com.puppycrawl.tools.checkstyle.internal.XdocsJavaDocsTest.testAllCheckSectionJavaDocs(XdocsJavaDocsTest.java:166)
```

```
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   XdocsPagesTest.testAllCheckSections:500->validateCheckSection:583->validatePropertySection:725->validatePropertySectionProperties:861->validatePropertySectionPropertyEx:892->getFieldClass:1260 config_coding.xml section 'OverloadMethodsDeclarationOrder' could not find field modifierGroups
expected to be true
[ERROR] Errors:
[ERROR]   XdocsJavaDocsTest.testAllCheckSectionJavaDocs:166->examineCheckSection:187->examineCheckSectionChildren:208->examineCheckSubSection:222->createPropertiesText:311 NullPointer
```

**Comment by rnveach** [comment]
@romani The examples in the issue do not compile and can't be copied to PR as it is.

````
> javac .\Test.java
.\Test.java:8: error: method foo() is already defined in class Test
        public static void foo() {
                           ^
1 error
````

**Comment by romani** [comment]
It ok, please update issue description name method to be unique and to be compilable if we want usage as is.

**Comment by petromir** [comment]
@romani Any plans to move this forward?

**Comment by romani** [comment]
everyone in web is welcome to send PR with fix.

**Comment by romani** [comment]
@Zopsss , do you have time to finish this issue ?

**Comment by Zopsss** [comment]
> @Zopsss , do you have time to finish this issue ?

Sure! I gone through the issue description and it seems interesting to work on, I'm currently working on few issues once I'll complete some of them I'll raise a PR for this one. Thanks for pointing out this :)

**Comment by austinarbor** [comment]
please note the description in the issue here is no longer accurate based on convo on my PR https://github.com/checkstyle/checkstyle/pull/12086#discussion_r1135345439

**Comment by Zopsss** [comment]
> please note the description in the issue here is no longer accurate based on convo on my PR https://github.com/checkstyle/checkstyle/pull/12086#discussion_r1135345439
> 
> there is also unanswered question https://github.com/checkstyle/checkstyle/pull/12086#discussion_r1174671496

Thanks for the information! Will keep this in mind 🤜🏻🤛🏻

**Comment by Zopsss** [comment]
Hey @romani I was thinking about starting working on this issue, I have cherry-picked the changes from @austinarbor's PR and also rebased my branch with master. I had few doubts I would like to clear before starting working on this issue, 

1. As you mentioned in this https://github.com/checkstyle/checkstyle/pull/12086#discussion_r1135345439 , We don't require modifiers to be in order right? So the following example should not generate any violations? Just asking for confirmation

```java
// ModifiersThatCreateGroup = "public static, static, public"
public class Test {
    void foo() { // default group
    }

    final void foo(String a, String b) { // default group
    }

    public static void foo(String a) { // first group
    }

    static public void foo(int a) { // first group
    }

    public void foo(double t) { // third group
    }

    static void foo(float x) { // second group
    }
}
```

2. As you said here https://github.com/checkstyle/checkstyle/pull/12086#discussion_r1155444486, 

> Method without modifiers will be separate group

This means `package` methods will have their separate group right ( let's say `package lvl` group )? And suppose if a method is declared as `final` or `abstract` or `static` etc but didn't specified any modifiers then that method will be counted in the `package lvl` group or it should be in it's own group? For Example following code should be fine:

```java
// ModifiersThatCreateGroup = "static private, protected abstract"
abstract public class Test {
    void foo() { // 'package lvl' group
    }

    final void foo(int a) { // 'package lvl' group
    }

    abstract void foo(int a, int b); // 'package lvl' group

    static void foo(String b) { // 'package lvl' group
    }

    private static void foo(double d) { // first group
    }

    protected abstract void foo(float x); // second group
}
```

3. If possible can you explain in detail how we should approach grouping of methods without any modifiers? Should we use the approach I mentioned in the 2nd point or we should try something else?

4. Is there something that I'm missing which is not specified in the issue description but later on got discussed?

**Comment by romani** [comment]
>We don't require modifiers to be in order right?

We do not require order of modifiers, user define modifiers separated by space. Explicit Modifiers from methods are used.

In your example you have 3 groups that overlap with each other. That should make some violations.

**Comment by Zopsss** [comment]
@romani ping

**Comment by Zopsss** [comment]
> In your example you have 3 groups that overlap with each other. That should make some violations.

can you also reply to my other points mentioned in the https://github.com/checkstyle/checkstyle/issues/3770#issuecomment-1975256555 ? ( point 2,3 & 4 )
